File: mint-back/app/schemas/company.py
from typing import Dict, List, Optional, Any
from pydantic import BaseModel, Field, ConfigDict, field_validator
from datetime import datetime
from app.models.task import TaskStatus, TaskType
from app.core.validators import InputValidator

class CompanyBase(BaseModel):
    name: str = Field(..., min_length=2, max_length=100, description="Company name")
    website: str = Field(..., min_length=4, max_length=255, description="Company website URL")
    
    @field_validator('name')
    @classmethod
    def validate_name(cls, v):
        """Validate and sanitize company name"""
        return InputValidator.validate_company_name(v)
    
    @field_validator('website')
    @classmethod
    def validate_website(cls, v):
        """Validate and sanitize website URL"""
        return InputValidator.validate_website_url(v)

class CompanyCreate(CompanyBase):
    pass  # Only inherits name and website from CompanyBase

class CompanyUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=2, max_length=100, description="Company name")
    website: Optional[str] = Field(None, min_length=4, max_length=255, description="Company website URL")
    profile: Optional[Dict[str, Any]] = Field(None, description="Company profile data")
    digital: Optional[Dict[str, Any]] = Field(None, description="Digital presence data")
    timeline: Optional[Dict[str, Any]] = Field(None, description="Company timeline data")
    products: Optional[Dict[str, Any]] = Field(None, description="Products data")
    jobs: Optional[Dict[str, Any]] = Field(None, description="Jobs data")
    csr: Optional[Dict[str, Any]] = Field(None, description="CSR data")
    press: Optional[Dict[str, Any]] = Field(None, description="Press data")
    team: Optional[List[Dict[str, Any]]] = Field(None, description="Team data")
    
    @field_validator('name')
    @classmethod
    def validate_name(cls, v):
        """Validate and sanitize company name"""
        if v is not None:
            return InputValidator.validate_company_name(v)
        return v
    
    @field_validator('website')
    @classmethod
    def validate_website(cls, v):
        """Validate and sanitize website URL"""
        if v is not None:
            return InputValidator.validate_website_url(v)
        return v
    
    @field_validator('profile', 'digital', 'timeline', 'products', 'jobs', 'csr', 'press')
    @classmethod
    def validate_json_fields(cls, v, info):
        """Validate JSON field data"""
        if v is not None:
            return InputValidator.validate_json_field(v, info.field_name)
        return v
    
    @field_validator('team')
    @classmethod
    def validate_team(cls, v):
        """Validate team data"""
        if v is not None:
            return InputValidator.validate_json_field(v, 'team')
        return v

class TaskResponse(BaseModel):
    id: int
    company_id: int
    type: TaskType
    status: TaskStatus
    error: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

class CompanyResponse(CompanyBase):
    id: int
    owner_username: str
    profile: Dict[str, Any] = Field(default_factory=dict)
    digital: Dict[str, Any] = Field(default_factory=dict)
    timeline: Dict[str, Any] = Field(default_factory=dict)
    products: Dict[str, Any] = Field(default_factory=dict)
    jobs: Dict[str, Any] = Field(default_factory=dict)
    csr: Dict[str, Any] = Field(default_factory=dict)
    press: Dict[str, Any] = Field(default_factory=dict)
    team: List[Dict[str, Any]] = Field(default_factory=list)
    error: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    tasks: List[TaskResponse] = Field(default_factory=list)
    
    model_config = ConfigDict(from_attributes=True) 

File: src/types/company.ts
import type { TaskResponse } from './task'

export interface CompanyCreate {
  name: string
  website: string
  owner_username: string
}

export interface CompanyUpdate {
  name?: string | null
  website?: string | null
  profile?: Record<string, any> | null
  digital?: Record<string, any> | null
  timeline?: Record<string, any> | null
  products?: Record<string, any> | null
  jobs?: Record<string, any> | null
  csr?: Record<string, any> | null
  press?: Record<string, any> | null
  team?: Record<string, any>[] | null
}

/**
 * Generic type for values with source attribution.
 * Matches backend SourcedValue[T] Pydantic schema.
 *
 * @template T - The type of the value field (string, number, array, object)
 *
 * @property value - The actual value of type T
 * @property source - Source URL or tool name (e.g., "Chaps-e", "mistral", URL)
 * @property favicon - Optional favicon URL for the source
 * @property value_fr - Optional French translation placeholder (future i18n support)
 */
export type SourcedValue<T> = {
  value: T
  source: string
  favicon?: string
  value_fr?: string
}

/**
 * Press item interface for normalized press data.
 * Uses singular 'source' field instead of the old 'sources[]' array.
 *
 * This change aligns with the backend database normalization where each
 * press item has a single source reference.
 */
export interface PressItem {
  value: string
  source: string
  value_fr?: string
}

export interface TeamMember {
  position: string
  firstName: string
  lastName: string
  linkedinUrl?: string
  subordinates?: TeamMember[]
}

export interface Company {
  id?: number
  pending?: boolean
  error?: string
  name: string
  website: string
  tasks: TaskResponse[]

  created_at: string
  updated_at: string
  created_by: string
  updated_by: string

  owner_username: string
  organization_id: string

  // Optional folder information (populated for recent companies)
  folder_id?: string
  folder_name?: string

  profile: {
    // Added insights field with SourcedValue pattern
    insights?: SourcedValue<string>
    groupName?: SourcedValue<string>
    businessLine?: SourcedValue<string>
    catchphrase?: SourcedValue<string>
    establishmentYear?: SourcedValue<string>
    employeeCount?: SourcedValue<string>
    revenue?: SourcedValue<string>
    ceo?: SourcedValue<string>
    hq?: SourcedValue<string>
  }

  digital: {
    insights?: string
    digitalStrategy?: SourcedValue<{
      overallStrategy: string
      digitalTransformation: string
      eCommerceCapabilities: string
      mobileStrategy: string
      digitalMarketingApproach: string
    }>
    onlineServices?: SourcedValue<
      {
        name: string
        description: string
      }[]
    >
    socialMediaAccounts?: {
      platform: string
      url: string
    }[]

    loyaltyProgram?: SourcedValue<string>
  }

  timeline: {
    insights?: string

    events?: {
      date: string
      title: string
      description: string
      category: string
      location: string
      impact: string
      source: string
    }[]
  }

  products?: {
    insights?: string
    customerType?: string
    marketingPositioning?: string
    range?: SourcedValue<string>[]
    partnerBrands?: SourcedValue<string>[]
    privateLabels?: SourcedValue<string>[]
    categories?: {
      [key: string]: string[]
    }
  }

  jobs: {
    offers?: {
      title: string
      location: string
      department: string
      description: string
      requirements: string
      posted_date: string
    }[]
    insights?: {
      total_openings: SourcedValue<number>
      top_departments: SourcedValue<string[]>
      hiring_focus: SourcedValue<string>
      growth_indicators: SourcedValue<string>
    }
  }

  csr: {
    insights?: string
    responsibility?: string
    responsibility_initiatives?: SourcedValue<string>[]
    charity_actions?: SourcedValue<string>[]
    sustainability_programs?: SourcedValue<string>[]
    community_involvement?: SourcedValue<string>[]
    diversity_inclusion?: SourcedValue<string>[]
    ethical_practices?: SourcedValue<string>[]
    awards_certifications?: SourcedValue<string>[]
  }

  /**
   * Press section with normalized press items.
   * BREAKING CHANGE: Press items now use singular 'source' field instead of 'sources[]' array.
   * This aligns with the backend database normalization.
   */
  press: {
    insights?: string
    articles?: PressItem[]
    press_releases?: PressItem[]
    media_mentions?: PressItem[]
    awards_recognition?: PressItem[]
    product_launches?: PressItem[]
    executive_interviews?: PressItem[]
    financial_news?: PressItem[]
    partnership_announcements?: PressItem[]
  }

  team?: TeamMember[]

  // Raw knowledge fields (debug/admin only)
  raw_mistral_knowledge?: string | null
  raw_claude_knowledge?: string | null
  raw_wikipedia_knowledge?: string | null
  raw_scraped_website_knowledge?: string | null
}
